exporter/xray/segment.go

```diff
@@ -0,0 +1,393 @@
+	}
+
+	var spanID trace.SpanID
+	copy(spanID[:], data)
+
+	return spanID, nil
+}
+
+// merge all string, bool, and numeric values from src to dest
+func merge(dest, src map[string]interface{}) {
+	for key, value := range src {
+		switch value.(type) {
+		case bool:
+			dest[key] = value
+		case string:
+			dest[key] = value
+		case int, int8, int16, int32, int64:
+			dest[key] = value
+		case uint, uint8, uint16, uint32, uint64:
+			dest[key] = value
+		case float32, float64:
+			dest[key] = value
+		}
+	}
+}
+
+func makeAnnotations(annotations []trace.Annotation, attributes map[string]interface{}) map[string]interface{} {
+	var result = map[string]interface{}{}
+
+	for _, annotation := range annotations {
+		merge(result, annotation.Attributes)
+	}
+	merge(result, attributes)
+
+	if len(result) == 0 {
+		return nil
+	}
+	return result
+}
+
+func makeCause(status trace.Status) (isError, isFault bool, cause *errCause) {
+	if status.Code == 0 {
+		return
+	}
+
+	if status.Message != "" {
+		id := make([]byte, 8)
+		mutex.Lock()
+		r.Read(id) // rand.Read always returns nil
+		mutex.Unlock()
+
+		hexID := hex.EncodeToString(id)
+
+		cause = &errCause{
+			Exceptions: []exception{
+				{
+					ID:      hexID,
+					Message: status.Message,
+				},
+			},
+		}
+
+		if dir, err := os.Getwd(); err == nil {
+			cause.WorkingDirectory = dir
+		}
+	}
+
+	if status.Code >= 400 && status.Code < 500 {
+		isError = true
+		return
+	}
+
+	isFault = true
+	return
+}
+
+func rawSegment(span *trace.SpanData) segment {
+	var (
+		traceID                 = MakeAmazonTraceID(span.TraceID)
+		startMicros             = span.StartTime.UnixNano() / int64(time.Microsecond)
+		startTime               = float64(startMicros) / 1e6
+		endMicros               = span.EndTime.UnixNano() / int64(time.Microsecond)
+		endTime                 = float64(endMicros) / 1e6
+		filtered, http, host    = makeHttp(span.Attributes)
+		isError, isFault, cause = makeCause(span.Status)
+		annotations             = makeAnnotations(span.Annotations, filtered)
+		name                    = span.Name
+		namespace               string
+	)
+
+	if host != "" {
+		name = host
+	}
+	if span.HasRemoteParent {
+		namespace = "remote"
+	}
+
+	return segment{
+		ID:          MakeAmazonSpanID(span.SpanID),
+		TraceID:     traceID,
+		Name:        name,
+		StartTime:   startTime,
+		EndTime:     endTime,
+		Namespace:   namespace,
+		ParentID:    MakeAmazonSpanID(span.ParentSpanID),
+		Annotations: annotations,
+		Http:        http,
+		Error:       isError,
+		Fault:       isFault,
+		Cause:       cause,
+	}
+}
+
+type writer struct {
+	buffer  *bytes.Buffer
+	encoder *json.Encoder
+}
+
+func (w *writer) Reset() {
+	w.buffer.Reset()
+}
+
+func (w *writer) Encode(v interface{}) error {
+	return w.encoder.Encode(v)
+}
+
+func (w *writer) String() string {
+	return w.buffer.String()
+}
+
+const (
+	maxBufSize = 256e3
+)
+
+var (
+	writers = &sync.Pool{
```

Comment:
It seems like aside from cleanup, these pooled objects will only ever be used from a single goroutine on a timer. If we could arrange for the cleanup to run on the same goroutine, then we might be able to avoid using a pool and simply keep a single buffer?

Reply:
Actually, looking closer it seems like access to the pool is always protected by e.mutex, so we probably can already do this.

Reply:
This routine is called from a number of places:

* on timer - to flush the queue
* when the buffer is full
* manually via Flush
* manually via Close

The primary consideration was when traces are being generated rapidly which would trigger to flush on full buffer case.

makeInput is only ever called by publish which always runs in its own goroutine.

That said, I don't feel strongly about this.  Happy to remove if you feel strongly.